bridges/snowbridge/parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The message is too large
+		MessageTooLarge,
+		/// The pallet is halted
+		Halted,
+		// Invalid fee config
+		InvalidFeeConfig,
+		/// Invalid Channel
+		InvalidChannel,
+	}
+
+	/// Messages to be committed in the current block. This storage value is killed in
+	/// `on_initialize`, so should never go into block PoV.
+	///
+	/// Is never read in the runtime, only by offchain message relayers.
+	///
+	/// Inspired by the `frame_system::Pallet::Events` storage value
+	#[pallet::storage]
+	#[pallet::unbounded]
+	pub(super) type Messages<T: Config> = StorageValue<_, Vec<CommittedMessage>, ValueQuery>;
+
+	/// Hashes of the ABI-encoded messages in the [`Messages`] storage value. Used to generate a
+	/// merkle root during `on_finalize`. This storage value is killed in
+	/// `on_initialize`, so should never go into block PoV.
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn message_leaves)]
+	pub(super) type MessageLeaves<T: Config> = StorageValue<_, Vec<H256>, ValueQuery>;
+
+	/// The current nonce for each message origin
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, ChannelId, u64, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		T::AccountId: AsRef<[u8]>,
+	{
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Remove storage from previous block
+			Messages::<T>::kill();
+			MessageLeaves::<T>::kill();
+			// Reserve some weight for the `on_finalize` handler
+			T::WeightInfo::commit()
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			Self::commit();
+		}
+
+		fn integrity_test() {
+			let decimals = T::Decimals::get();
+			assert!(decimals == 10 || decimals == 12, "Decimals should be 10 or 12");
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Halt or resume all pallet operations. May only be called by root.
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			mode: BasicOperatingMode,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			OperatingMode::<T>::put(mode);
+			Self::deposit_event(Event::OperatingModeChanged { mode });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Generate a messages commitment and insert it into the header digest
+		pub(crate) fn commit() {
+			let count = MessageLeaves::<T>::decode_len().unwrap_or_default() as u64;
+			if count == 0 {
+				return
+			}
+
+			// Create merkle root of messages
+			let root = merkle_root::<<T as Config>::Hashing, _>(MessageLeaves::<T>::stream_iter());
+
+			// Insert merkle root into the header digest
+			<frame_system::Pallet<T>>::deposit_log(sp_runtime::DigestItem::Other(
```

Comment:
Maybe it makes sense to add something similar to:
```rust
enum CustomDigestItem {
    /// Merkle root of outbound Snowbridge messages.
   SnowbridgeOutbound(Vec<u8>),
}
...
deposit_log(sp_runtime::DigestItem::Other(CustomDigestItem::SnowbridgeOutbound(root.to_fixed_bytes().into())));
```

Who knows what will happen later with this chain - maybe we add some other bridges, who will be using their own digest items?

I know that this code doesn't belong to this repo and has been reviewed and audited elsewhere - just stumbled upon it :) Feel free to ignore

Reply:
(I mean `CustomDigestItem` should be added to runtime and there must be some pallet config parameter to implement the wrapping. Also by "this chain" I mean all our bridge hubs)

Reply:
Ok, I see there's a dependency on that in the Solidity contract [here](https://github.com/Snowfork/snowbridge/blob/c03175b813ee97373843de4f70ea10df1977beb7/contracts/src/Verification.sol#L128-L141). So maybe it is too late already for this change...Still we need to be super careful in the future if we'll be customizing BH digests (i.e. adding other bridges) @paritytech/bridges-core

Reply:
I totally agree - I've known about this issue, though it slipped through the cracks, as it was fairly low priority. Its also something that can be remedied in a post-launch upgrade of BridgeHub and our gateway contracts.

Nevertheless, I will look into fixing it this week, since its a fairly trivial fix on both sides of the bridge

Reply:
OK, I've added a `CustomDigestItem` type in this PR: https://github.com/Snowfork/polkadot-sdk/pull/65/files#diff-4e5ae0ed65c739f7157ca147180243eadb8fbaf42a74efdb2a8c6466662597e3
However, I've retained the existing encoding, which helps reduce gas costs on Ethereum.

Let me know what you think...

Reply:
I think it is cool! Thank you!